### PR TITLE
#155 258 145 issues with main

### PR DESCRIPTION
**Pull request by Mares2022**
The new pull request integrates the calculation of ground elevation from a digital elevation file (.tiff), the tool is flexible and calculates Ground Elevation just in case a dem file is provided. Currently .setup_buildings_from_single_source() is the only option that has been tested. The function was also integrated in .setup_buildings_from_multiple_sources() but I require test data that triggers this function. In .setup_buildings_from_single_source() I have realized that it is difficult to access the geometries (which is required in setup_ground_elevation), even after running .set_exposure_geoms() which gives me the impression that this function is not working correctly, because it creates a list of geometries and not a geopandas dataframe (but maybe that is the way this function works). I accessed to the geometries maybe in to the best optimal way so is best to double check if that is the best alternative.

### CODE REVIEW

**Comment by Mares2022** [comment]
I have in the branch all the outputs of my tests, for one reason I couldn't add them in git ignore file without creating conflicting the .gitignore of the main branch

File: tests/test_setup_new_composite_areas_ground_elevation.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    # read modified exposure
+    exposure_modified = pd.read_csv(
+        _cases[case]["new_root"] / "exposure" / "exposure.csv"
+    )
+
+    exposure_modified = fm.exposure.exposure_db
+
+    # check if the new development area was added
+    assert len(exposure_modified) > len(exposure_original), 'The aggregated areas were not added'
```
Comment (frederique-hub) [inline tests/test_setup_new_composite_areas_ground_elevation.py:89]: Maybe you can also add here the check that the Ground Elevation column is present in the exposure.csv. You can do that with for example

`assert "Ground Elevation" in fm.exposure.exposure_db.columns`

and

`assert fm.exposure.exposure_db["Ground Elevation"].isna().sum() == 0`

File: hydromt_fiat/workflows/exposure_vector.py [hunk 2]
```diff
@@ -538,52 +548,89 @@ def setup_ground_floor_height(
 
     def setup_max_potential_damage(
         self,
-        max_potential_damage: Union[List[float], float],
-        damage_types: Union[List[str], str, None] = None,
-        country: Union[str, None] = None,
-    ):
-        if max_potential_damage == "jrc_damage_values":
```
Comment (frederique-hub) [inline hydromt_fiat/workflows/exposure_vector.py:545]: Please keep the part from line 545-550 and 553-557 in this function, that you removed. They are required when people use hazard or JRC as a source for the max potential damages. Your part should be added on top; so making more options possible.

File: hydromt_fiat/workflows/exposure_vector.py [hunk 3]
```diff
@@ -538,52 +548,89 @@ def setup_ground_floor_height(
...
-        if damage_types is None:
-            damage_types = ["total"]
+    def setup_ground_elevation(
+        self,
+        ground_elevation: Union[int, float, None, str, Path, List[str], List[Path]],
+        exposure_db: pd.DataFrame,
+        exposure_geoms: gpd.GeoDataFrame,
+    ) -> None:
 
-        for damage_type in damage_types:
```
Comment (frederique-hub) [inline hydromt_fiat/workflows/exposure_vector.py:571]: Lines 571-586 should also be included still I think. Please check how the `setup_max_potential_damage` function is now in the main branch and add your code then as other if/else options

File: hydromt_fiat/workflows/exposure_vector.py [hunk 4]
```diff
@@ -844,7 +892,7 @@ def setup_new_composite_areas(
 
         percent_growth = float(percent_growth) / 100
         geom_file = Path(geom_file)
-        assert geom_file.is_file()
+        assert geom_file.is_file() #TODO: The assert message is missing
```
Comment (frederique-hub) [inline hydromt_fiat/workflows/exposure_vector.py:895]:
```suggestion
        assert geom_file.is_file() f"File {str(geom_file)} is missing, cannot set up a new composite area."
```

File: hydromt_fiat/workflows/exposure_vector.py [hunk 5]
```diff
@@ -966,6 +1014,9 @@ def setup_new_composite_areas(
                 self.crs,
             )
 
+        # Adding elevation data into the new objects
+        self.setup_ground_elevation(ground_elevation,new_objects,_new_exposure_geoms)
```
Comment (frederique-hub) [inline hydromt_fiat/workflows/exposure_vector.py:1018]: Very tiny thing but we can do formatting with Ruff by running `ruff .` in the main folder of the repository.
```suggestion
        self.setup_ground_elevation(ground_elevation, new_objects, _new_exposure_geoms)
```

File: tests/test_ground_elevation.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    # Remove the new root folder if it already exists
+    if _cases[case]["new_root"].exists():
+        shutil.rmtree(_cases[case]["new_root"])
+
+    # Set the new root and write the model
+    fm.set_root(_cases[case]["new_root"])
+    fm.write()
+
+    unique_ge_new = fm.exposure.exposure_db["Ground Elevation"].unique()
```
Comment (frederique-hub) [inline tests/test_ground_elevation.py:55]: Is this the most logical test to do? Maybe good to think more about how to assess the Ground Elevation column is filled. Maybe it cannot be checked by values but that all values are filled with a reasonable number maybe is good?

File: tests/test_integrations_hazard.py
Comment (frederique-hub) [inline tests/test_integrations_hazard.py:1]: I like these improvement!

File: tests/test_update_max_potential_damage_user_data.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+EXAMPLEDIR = Path().absolute() / "examples" / "data" / "update_max_potential_damage"
+DATADIR = Path().absolute() / "hydromt_fiat" / "data"
+
+_cases = {
+    "update_max_potential_damage_with_points": {
+        "dir": "fiat_model",
+        "new_root": EXAMPLEDIR / "test_update_max_potential_damage_points",
+        "data_catalog": DATADIR / "hydromt_fiat_catalog_USA.yml",
+        "max_potential_damage_file": EXAMPLEDIR
```
Comment (frederique-hub) [inline tests/test_update_max_potential_damage_user_data.py:18]: You're using the same geometry files for this test as for the update_ground_floor_height test. Is that the idea? Maybe it does not matter for the testing itself but the naming is confusing.

File: tests/test_update_max_potential_damage_user_data.py
Comment (frederique-hub) [inline tests/test_update_max_potential_damage_user_data.py:1]: Did you also check the outputs by hand? That the right values are taking over by the correct building points?

File: hydromt_fiat/workflows/exposure_vector.py [hunk 8]
```diff
@@ -538,52 +548,89 @@ def setup_ground_floor_height(
...
-            except KeyError as e:
-                self.logger.warning(
-                    f"Not found in the {max_potential_damage} damage "
-                    f"value data: {e}"
-                )
+        if ground_elevation:
+            # This function was developed by Willem https://github.com/interTwin-eu/DT-flood/blob/DemonstrationNotebooks/Notebooks/SetupFIAT.ipynb
+            #TODO: Find equivalent functions in hydromt
+            # Read in the DEM
+            dem = rasterio.open(ground_elevation)
```
Comment (frederique-hub) [inline hydromt_fiat/workflows/exposure_vector.py:596]: Can you put the code under `if ground_elevation:` for getting the ground elevation and adding it to the exposure objects in a separate function in gis.py?

**Review by frederique-hub — CHANGES_REQUESTED** [review]
Hi Mario, thanks for all the work and it is quite good already. There are a few medium effort items but most are small things. Could you let me know when you have time to work on this? Thank you!

File: envs/hydromt-fiat-dev.yml
[the remark this answers is not in the document]
Comment (Mares2022) [reply envs/hydromt-fiat-dev.yml:1]: I included them in the environment hydromt-fiat-dev.yml file, and the project.toml file

File: tests/test_setup_exposure_buildings_with_dem_data.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+                "asset_locations": "NSI",
+                "occupancy_type": "NSI",
+                "max_potential_damage": "NSI",
+                "damage_types": ["structure", "content"],
+                "ground_floor_height": "NSI",
+                "unit": "ft",
+            },
+        },
+        "region": _region,
+    },
```
[the remark this answers is not in the document]
Comment (Mares2022) [reply tests/test_setup_exposure_buildings_with_dem_data.py]: I removed it and changed the name of the test with test_vulnerability_expousure_with_user_dem_data.py

File: tests/test_setup_exposure_buildings_with_dem_data.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    fm = FiatModel(root=root, mode="w", data_libs=[data_catalog_yml], logger=logger)
+
+    region = gpd.GeoDataFrame.from_features(_cases[case]["region"], crs=4326)
+    fm.build(region={"geom": region}, opt=_cases[case]["configuration"])
+    fm.write()
+
+    # Check if the exposure data exists
+    assert root.joinpath("exposure", "buildings.gpkg").exists()
+    assert root.joinpath("exposure", "exposure.csv").exists()
+    assert root.joinpath("exposure", "region.gpkg").exists()
```
[the remark this answers is not in the document]
Comment (Mares2022) [reply tests/test_setup_exposure_buildings_with_dem_data.py]: I included two new asserts to evaluate if the column has no null values, and that the values are within a maximum and minimum values of the raster file

File: tests/test_ground_elevation.py (hunk 6 above)
Comment (Mares2022) [reply tests/test_ground_elevation.py:58]: I included the same asserts than test_setup_exposure_buildings_with_dem_data.py test

File: tests/test_update_max_potential_damage_user_data.py (hunk 7 above)
Comment (Mares2022) [reply tests/test_update_max_potential_damage_user_data.py:18]: I modified the files names to make them clearer

File: tests/test_update_max_potential_damage_user_data.py
Comment (Mares2022) [reply tests/test_update_max_potential_damage_user_data.py:1]: I miss this components, should I check it by hand or try to make an assertion out of it?

File: hydromt_fiat/workflows/exposure_vector.py (hunk 8 above)
Comment (Mares2022) [reply hydromt_fiat/workflows/exposure_vector.py]: It is implemented now in gis.py

File: tests/test_update_max_potential_damage_user_data.py
Comment (Mares2022) [reply tests/test_update_max_potential_damage_user_data.py:1]: I miss this components, should I check it by hand or try to make an assertion out of it?

**Review by frederique-hub — APPROVED** [review]

**Pull request merged**